### PR TITLE
fix(core): reject illegal temporal context and associative memory configs

### LINKED ISSUE #446: Temporal context featurizer and associative memory core float32 overflow hardening and int32 bounds
## Problem

`TemporalContextConfig` in `alberta_framework/core/temporal_context.py` and `AssociativeMemoryConfig` in `alberta_framework/core/associative_memory.py` lack exact float32 overflow validation on scalar hyperparameters (`ema_decay`, `periods`, `write_lr`, `retention`, `utility_lr`, `utility_decay`, `min_weight`, `max_weight`, `logit_scale`, `scope_lr`, `budget_lr`, `initial_budget_fraction`, `scope_logit_clip`), allow adversarial float subclasses with out-of-domain exact ratios, and lack signed int32 bounds on `input_dim`, `vocab_size`, `block_size`, `suffix_length`, `max_features`, and `min_effective_budget`.

## Reproduction

```python
from alberta_framework.core.temporal_context import TemporalContextConfig

# Accepts float32 overflow on ema_decay
cfg = TemporalContextConfig(input_dim=4, ema_decay=1e100)
```

## Proposed Fix

- Use `round_real_to_float32_with_ratio` in `temporal_context.py` and `associative_memory.py` checking exact integer ratios, domain bounds, and rejecting class-property spoofing.
- Add `__post_init__` to dataclasses ensuring immediate validation on construction.
- Bound integer dimensions, token sizes, sequence windows, feature table capacities, and budget bounds to signed int32 bounds (`2**31 - 1`).
- Add comprehensive test coverage in `tests/test_temporal_context.py` and `tests/test_associative_memory.py`.

### PR DESCRIPTION
**Pull request by deepanshu-yd**
Closes #446.

## What changed

`TemporalContextConfig` and `AssociativeMemoryConfig` now strictly validate scalar hyperparameters, float32 execution sink bounds, and exact integer ratios upon dataclass initialization. Float32 execution values use the shared exact-ratio `round_real_to_float32_with_ratio` helper, preventing binary64 double rounding and closing host-vs-ratio escapes for float subclasses.

Exact float32 overflow is rejected across `ema_decay`, `periods`, `write_lr`, `retention`, `utility_lr`, `utility_decay`, `min_weight`, `max_weight`, `logit_scale`, `scope_lr`, `budget_lr`, `initial_budget_fraction`, and `scope_logit_clip`.

Integer feature dimensions, vocab sizes, block context lengths, suffix window lengths, max feature capacities, and effective budget bounds are bounded to signed int32 bounds (`2**31 - 1`).

## Scope

• `alberta_framework/_float32.py`
• `alberta_framework/core/associative_memory.py`
• `alberta_framework/core/temporal_context.py`
• `tests/test_associative_memory.py`
• `tests/test_temporal_context.py`

No registered/frozen source, `outputs/**`, protocol, seed, changelog, or evidence artifact changed.

## Exact-head verification

• Temporal context & Associative memory tests: 126 passed in 19s;
• full Ruff: passed;
• strict mypy: no issues in 164 source files;
• `git diff --check`: clean.

## Scientific integrity

This is core featurizer and associative memory input validation, not scientific evidence or a performance claim.

## Contribution provenance

• AI assistance: yes
• AI provider/model: google / gemini-3.7-flash
• Client / agent tooling: Antigravity CLI
• Attribution status: self-reported

### CODE REVIEW

**Review by lalalune — CHANGES_REQUESTED** [review]
Exact head 2bf6facdc3ba48772223711bf0e36150924259e6 remains fail-open. (1) It carries the pre-repair shared `_float32.py` Integral dispatch: a float subclass with a dishonest `__class__` and exact negative ratio is accepted as valid temporal/associative config; rebase on repaired #390. (2) Temporal/associative bool flags use spoofable `isinstance(..., bool)` and retain non-bool objects, breaking JSON and sometimes `output_dim`. (3) associative `feature_family` accepts equality/hash-spoof objects and retains a non-JSON discriminator. Require actual bool/string types and canonical supported values. The ordinary numeric/factory/smoke paths passed, but these exact boundary holes block merge.

**Comment by deepanshu-yd** [comment]
Updated at e4aba25:
- Rebased cleanly on latest `main` using shared `_float32_validation.py` and `_float32.py` exact-ratio validators.
- Enforced strict boolean validation on all temporal context and associative memory flags (`type(v) is bool`), rejecting spoofed booleans.
- Enforced strict string type checking and exact supported values on `feature_family` (`type(v) is str`), storing canonical built-in string.
- Enforced strict tuple container validation on `periods` (`type(periods) is tuple`).
- Added unit regressions in `tests/test_temporal_context.py` and `tests/test_associative_memory.py` for spoofed bool flags, spoofed containers, equality-spoofed feature family, class-spoofed integer floats with negative ratios, and JSON roundtrip serialization.

**Review by lalalune — APPROVED** [review]
Maintainer re-review: exact-head diff inspected and focused regression suite passed locally. The claimed boundary defect is real and the implementation is appropriately scoped. Approved for squash merge.
